data_structures.hpp

```diff
@@ -125,75 +131,96 @@ void list<T>::delete_last(){                   //удаляет последни
 
 template <typename T>
 void list<T>::delete_concrete(T _value){       //удаляет все элементы с заданным значением
-    if(is_empty()){                         //проверка списка на наличие элементов
+    if(is_empty()){
         return;
     }
-    while(first -> value == _value){        //проверка равенства значения первого элемента списка значению _value (зачем нам вытаскивать value если first и first -> value это одно и тоже)
+    while(first -> value == _value){
         delete_first();
     }
-    /*else if(last -> value == _value){     //проверка равенства значения последнего элемента списка значению _value
-        delete_last();
-        return;
-    }*/
     if(!is_empty()){
-        Node<T>* slow = first;                      //создаем указатель на структуру типа Node и присваиваем значение указателя first
-        Node<T>* fast = slow -> next;               //создаем указатель на структуру типа Node и присваиваем значение next указателя slow
-        while(fast){                                //проверка указателя fast и значения value на соответсвие значение _value                 
+        Node<T>* slow = first;
+        Node<T>* fast = slow -> next;
+        while(fast){       
             if(fast -> value == _value){
                 slow -> next = fast -> next;
                 delete fast;
                 fast = slow -> next;
             }
             else{
                 slow = fast;
-                fast = fast -> next; 
-            /*else if(fast == _value){
-                delete fast;
-            }*/ 
+                fast = fast -> next;
             }                      
         }
     }
 }
 
 template <typename T>
-T& list<T>::operator[](const int num){                     //требуется объяснение         добавляет оператор "[]"
+T& list<T>::operator[](const int num){                     //добавляет оператор "[]"
     Node<T>* p = first;
     for(int i = 0; i < num && p; i++){
         p = p -> next;
     }
-    T& ret_val = p -> value;
-    return ret_val;
+    T& rval = p -> value;
+    return rval;
+}
+/*
+template<typename T>
+list<T>& list<T>::operator=(const list<T>& copy){
+    if(this == &copy){
+        return *this;
+    }
+    Node<T>* p = first;
+    while(p){
+        p = p -> next;
+        delete first;
+        first = p;
+    }
+    p = copy.first;
+    while(p){
+        this -> push_back(p -> value);
+        p = p -> next;
+    }
+    return *this;
+}*/
+
+template<typename T>
+list<T>& list<T>::operator=(const list<T>& copy){
```

Comment:
Since you expect having copy constructor in `T` we can make limitation to have copy assignment as well. In this case algorithm may be simplified.